pynucastro/rates/library.py

```diff
@@ -124,28 +122,118 @@ class Library:
 
     def __init__(self, libfile=None, rates=None):
         self._library_file = libfile
+        self._rates = {}
+
         if rates:
-            self._rates = None
             if isinstance(rates, Rate):
                 rates = [rates]
             if isinstance(rates, dict):
                 self._rates = rates
             elif isinstance(rates, (list, set)):
-                self._add_from_rate_list(rates)
+                self.add_rates(rates)
             else:
-                raise TypeError("rates in Library constructor must be a Rate object, list of Rate objects, or dictionary of Rate objects keyed by Rate.get_rate_id()")
-        else:
-            self._rates = {}
+                raise TypeError("rates in Library constructor must be a Rate object, list of Rate objects, or dictionary of Rate objects keyed by Rate.id")
+
         self._library_source_lines = collections.deque()
 
         if self._library_file:
             self._library_file = _find_rate_file(self._library_file)
             self._read_library_file()
 
+    def get_rates(self):
+        """ Return a list of the rates in this library."""
+        return list(self._rates.values())
+
+    def get_rate(self, rid):
+        """ Return a rate matching the id provided. """
+        try:
+            rid_mod = capitalize_rid(rid, " ")
+            return self._rates[rid_mod]
+        except KeyError:
+            pass
+
+        # fallback to the rate fname
+        try:
+            rid_mod = capitalize_rid(rid, "_")
+            return [q for q in self.get_rates() if q.fname == rid_mod][0]
+        except IndexError:
+            raise LookupError(f"rate identifier {rid!r} does not match a rate in this library.") from None
+
+    @property
+    def num_rates(self):
+        return len(self.get_rates())
+
+    def add_rate(self, rate):
+        """Manually add a rate by giving a Rate object"""
+
+        try:
+            rid = rate.id
+        except AttributeError as exc:
+            raise TypeError(f"invalid Rate object {rate}") from exc
+
+        if rid not in self._rates:
+            self._rates[rid] = rate
+
+    def add_rates(self, ratelist):
+        """ Add to the rate dictionary from the supplied list of Rate objects."""
+
+        for rate in ratelist:
+            if rate.id in self._rates:
+                raise ValueError(f"supplied a Rate object already in the Library: {rate.id}")
```

Comment:
I think `add_rate()` should raise this error too, instead of silently doing nothing. @zingale?

Reply:
agreed